src/main/java/it/gov/pagopa/pu/debtpositions/service/statusalign/StatusRulesHandler.java

```diff
@@ -0,0 +1,87 @@
+package it.gov.pagopa.pu.debtpositions.service.statusalign;
+
+import java.util.List;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+public abstract class StatusRulesHandler<E extends Enum<E>> {
+
+  private final E syncStatus;
+  private final E paidStatus;
+  private final E unpaidStatus;
+  private final E expiredStatus;
+  private final E cancelledStatus;
+  private final E reportedStatus;
+  private final E invalidStatus;
+
+  protected StatusRulesHandler(E syncStatus, E paidStatus, E unpaidStatus, E expiredStatus, E cancelledStatus, E reportedStatus, E invalidStatus) {
+    this.syncStatus = syncStatus;
+    this.paidStatus = paidStatus;
+    this.unpaidStatus = unpaidStatus;
+    this.expiredStatus = expiredStatus;
+    this.cancelledStatus = cancelledStatus;
+    this.reportedStatus = reportedStatus;
+    this.invalidStatus = invalidStatus;
+  }
+
+  public <T, S, C extends Enum<C>> void updateEntityStatus(T entity,
+                                                           Function<T, List<S>> childStatusExtractor,
+                                                           Function<List<S>, C> statusDeterminer,
+                                                           BiConsumer<T, C> statusUpdater,
+                                                           BiConsumer<T, C> repositoryUpdater) {
+
+    List<S> childStatuses = childStatusExtractor.apply(entity);
+    C newStatus = statusDeterminer.apply(childStatuses);
+    statusUpdater.accept(entity, newStatus);
+    repositoryUpdater.accept(entity, newStatus);
+  }
+
+  public boolean isToSync(List<E> statusList) {
+    return statusList.contains(syncStatus);
+  }
+
+  public boolean isPartiallyPaid(List<E> statusList) {
+    return statusList.contains(paidStatus) &&
+      (statusList.contains(unpaidStatus) || statusList.contains(expiredStatus));
+  }
+
+  public boolean isUnpaid(List<E> statusList) {
+    return allMatchOrEmpty(statusList, status -> status.equals(unpaidStatus)) ||
+      (allMatchOrEmpty(statusList, status -> status.equals(cancelledStatus) || status.equals(unpaidStatus)) &&
+        statusList.contains(unpaidStatus));
```

Comment:
This is not efficient, you are iterating statusList for each status!

what you should implement is:
`statusList.contains(unpaidStatus) && statusList.stream().allMatch(status -> status.equals(cancelledStatus) || status.equals(unpaidStatus));`

you could generalize it the following way:
```
private allMatch(List<E> statusList, E requiredState, Set<E> allowedStatuses){
statusList.contains(requiredState) && statusList.stream().allMatch(status -> requiredState.equals(status) || allowedStatuses.contains(status));
}
```